### PR TITLE
"inner" index reorganization

### PR DESCRIPTION
**Pull request by tacaswell**
The top bar is too crowded so this moves the Third-party and contributing links out of the top bar and moves them to the version-specific landing page.

### CODE REVIEW

**Comment by story645** [comment]
Agree it's too cluttered, but what about moving reference to the version specific pages since it only makes sense in context of a specific version?
Maybe third-party-package should be renamed ecosystem since that's what everyone else is calling their analogous pages?

**Comment by timhoffm** [comment]
Agreed that it's too crowded. However, I'm neither happy with the current nor with the proposed structure. I'd need a moment to sit down and work out what's really the issue and propose something better. In the context of the upcoming 3.5 release: When do we need this to be finished? Are there plans to still work on the index page?

**Comment by tacaswell** [comment]
I think we have 2-3 weeks to get the inner index (as produced by sphinx) and the brochure site (@dorafc is looking at it right now) ready.

----

Doing a quick survey, the headings across the top for :

numpy: ['User Guide', 'API reference', 'Development']  https://numpy.org/doc/stable/index.html

pandas: ['Getting Started', 'User Guide', 'API reference', 'Development', 'Release notes'] https://pandas.pydata.org/docs/getting_started/index.html

scipy: ['Getting started', 'User Guide', 'API refernec', 'Development', 'Release notes'] http://scipy.github.io/devdocs/index.html

networkx: ['Install', 'Tutorial', 'Reference', 'Developer', 'Gallery', 'Guides->'] https://networkx.org/documentation/stable/index.html

Bokeh: [<version drop down>, 'First steps', 'User Guide', 'Gallary', 'Reference', 'Developers', 'Releases', 'Tutorial->', 'Community->']  https://docs.bokeh.org/en/latest/index.html

binder: ['Tutorials', 'How-to Guides', 'Reference', 'About mybinder.org', 'Contribute']  https://mybinder.readthedocs.io/en/latest/
----

For one thing the consensus seems to be to spell it "User Guide" not "User's Guide".  I have not chased down what is in each of those 

---

I'm going to push a bunch more re-organization work to this branch (some of it conceptual, some of it trying to get the navigation to play nice). 

I think putting the release notes an their own top-level thing makes a lot of sense.  My current plan is:

- pull "installation" out of the top-bar
- put the `pip install matplotlib` and `conda install matplotlib` options as the first section of the inner index page
- move the detailed installation to the "User Guide"
- put the release notes in the top bar
- put the learn / reference / how-to / understand into the actual grid. 
- move the "back matter" (citing, license, history, credit) to be at the bottom of the inner index page

---

The user profiles I'm thinking about are and what they want from the site in order to make progress on what ever it is they are actually doing that they need to use Matplotlib on the path to.

 - new-to-Matplotlib users (who have already decided / been told they are going to use Matplotlib).  They need to find the installation instructions (including links to when the happy-path does not work), the "plot types", the tutorials, the gallery, and where to interact with a human if they want to (probably in that order).  While installation is important, it is a one(few)-time thing (and we have the same process as everyone else on the standard platforms at this point) so I do not think we need to spend the real-estate on every single inner page on a direct link to installation, prominent placement on the inner index page is enough.  
 - current Matplotlib users.  This is people who have read some examples on line or gone through a boot camp.  They have used Matplotlib to get their work done in the past, they have come back to us with a specific goal of the visualization they want to make.  They are probably just making visualizations for themselves (or to present only the output to others) and most of the code they write using Matplotlib is going to be one-offs in a terminal or in a jupyter notebook.   The Plot Types sub-gallery is not as useful to them (they _know_ we can do bar charts, but now they want to do stacked bars with labels and error bars on them).  We need to direct these users to the gallery, the API reference, the tutorials, and then channels to talk to a human.
 - Matplotlib "power users".  This is people who are extensive users of Matplotlib and have written significant amounts of code that depends on Matplotlib that is in turn used by other people or users who are trying to push the boundaries of what you can do with Matplotlib.  They probably want access to the API reference, the gallery, the user guide, the release notes, the tutorials, and then channels to talk to  a human.

**Comment by story645** [comment]
> However, I'm neither happy with the current nor with the proposed structure. I'd need a moment to sit down and work out what's really the issue and propose something better

I think we all agree it's trying to do too many things, and it seems like Tom's proposal is slim it down to just folks who need to read the docs (judging by the user storys), and let the new "matplotlib.org" handle the community roadmapping (ecosystem & contributing). 

Plot Types and Gallery are also overlapping in weird ways, but I dunno that a gallery landing page/drop down is the right solution/too much. 

I'm unsure about `user guide` as a top level link 'cause currently it isn't and instead acts as a half filled index. I think Tom is trying to address some of this by moving installation there, and reworking how the things in the guide are linked,  but the current contents of the guide are basically all things that folks have proposed should be moved:
* interactive figures & fonts -  #19688 move to tutorial
* release notes - this should have it's own page and TOC, not be enumerated in the guide like this
* license/citation/credits/history - move links to TOC & lives in brochure site
* external resources - https://github.com/matplotlib/mpl-third-party/issues/71
* FAQ - #17920

File: doc/index.rst
Comment (timhoffm) [inline doc/index.rst:62]: Reference is a learning resource. Maybe "User documentation" as all this information is directed at our users. Unfortunately, this is close to the existing "User Guide", but the problem is rather that that is a weird catch-all and should probably be renamed (but even if we keep "User Guide" for now, that should not stop us from using "User documentation" here).

File: doc/users/resource.rst
Comment (timhoffm) [inline doc/users/resource.rst:1]: Make this two separate elements? It's a bit awkward to have them grouped together because, they don't have anything to do with each other.

File: doc/users/release_notes.rst
Comment (timhoffm) [inline doc/users/release_notes.rst:12]: What is the benefit of splitting this page? AFAICS there is no difference in the rendering other than we have additional but pretty useless toc pages per version (https://63048-1385122-gh.circle-artifacts.com/0/doc/build/html/users/relnotes/3.4.html).

From a maintainance POV I think having everything in one file is simpler because one can more easily see what is there.

File: doc/index.rst
Comment (timhoffm) [inline doc/index.rst:108]:
```suggestion
Third-party packages
```

https://63048-1385122-gh.circle-artifacts.com/0/doc/build/html/devel/documenting_mpl.html#section-name-formatting

Also in a number of other headings in here.

**Review by timhoffm — COMMENTED** [review]
Generally, this is on a good way. I particularly like the landing page.

I still struggle with the toc structure. The landing page is great for topical redirection, but we don't have a toc anywhere that shows everything we have. E.g. Contributing is only linked at the bottom of the landing page.

Since we cannot use the top bar as top-level navigation, we need a separate page that structures everything and contains the full TOC. For now, I call this "Site map", but there may be a better name (maybe just "Matplotlib 3.5 documentation"?). Structure could be:

- Site map
  - Installation
  - User documentation
    - Explanations
    [ can gallery and tutorials be linked in here? ]
    - API reference
    - FAQ
  - External resources
  - Release notes
  - Back matter
    - License
    - Citing
    - Credits
    - History
  - Contributing

The landing page  (index.rst) should not define the structure. It's not meant as a full overview and thus cannot be the top-level defining element. It rather announces the most relevant parts of the docs and links there.

File: doc/index.rst
Comment (story645) [inline doc/index.rst:72]:
```suggestion
    - :doc:`Plot Types <plot_types/index>`
```
it's an enumeration of the plot types Matplotlib supports

File: doc/index.rst
Comment (story645) [inline doc/index.rst:103]: the indent renders with two much white space, so wondering if there's a different way to make that subheading clear, or if this should be something like 
* pyplot 
    * create figures pyplot.figure
    * create subplots (pyplot.subplots, pyplot.subplot_mosaic) 
* axes
* figure
* rest of API
* toolkits

though wondering if there's a way to structure this to emphasize OO
image

File: doc/index.rst
Comment (story645) [inline doc/index.rst:89]: I think a show not tell approach could work better here, something like 
```suggestion
    -  :doc:`Tutorials <tutorials/index>` on topics such as usage, customization, color, fonts, etc. 
```

File: doc/index.rst
Comment (story645) [inline doc/index.rst:121]:
```suggestion
   :maxdepth: 1
```
 on this page, the whole contributing TOC tree is a bit distracting/overwhelming & this might be better as a small section w/ some copy on why folks should contribute or possibly stealing from the contribute section
> Thank you for your interest in helping to improve Matplotlib! There are various ways to contribute to Matplotlib, learn more at <devel/index>

File: doc/users/backmatter.rst
Comment (story645) [inline doc/users/backmatter.rst:2]:
```suggestion
About Matplotlib
```

**Review by story645 — COMMENTED** [review]
I still don't like that the user guide is functioning more as a site map than a guide - who is the intended audience for the user guide?

**Comment by timhoffm** [comment]
> I still don't like that the user guide is functioning more as a site map than a guide - who is the intended audience for the user guide?

That is addressed in my comment https://github.com/matplotlib/matplotlib/pull/20867#pullrequestreview-735615839. The user guide is gone there. I don't think we need a section "user guide" at all.

File: doc/users/release_notes.rst
Comment (tacaswell) [reply doc/users/release_notes.rst:12]: Before doing this re-org the list on the left has nice accordion things 
image
 instead of being one long list on my local build.  I'm using the latest pydata-sphinx-theme which apparently renders differently....

File: doc/users/resource.rst
Comment (tacaswell) [reply doc/users/resource.rst:1]: They are sort of lost-ducklings that ended up together....

File: doc/index.rst
Comment (tacaswell) [reply doc/index.rst:108]: 🐑

**Comment by tacaswell** [comment]
> I still struggle with the toc structure. 

It is made worse because the `.. toc::` structure is also used by sphinx to decide things like the "next" buttons, the auto-tocs on the left and right, and the the bread-crumbs in the top-bar (for example if you put anything else from the top-bar in the user in another one, they both show bold when you are in the inner one which I found very confusing). 

https://63048-1385122-gh.circle-artifacts.com/0/doc/build/html/contents.html does existing but we do not link to it anyplace obvious.  I guessed that it existed, because a contents.rst exists and suppressing sphinx from making a page while also using it to populate the top bar seemed hard! From there you can search for it and we do link to it a few random places (like the prose at the top of gallery).  

I _think_ the toc rules should be that `contents.rst` is the root of all files that have tocs and you can only use `.. toc::` in files that are in a toc in a file that is already referenced. 

My concern with your proposed layout @timhoffm is that we lose a way to pull apart the 4 kinds of docs that https://documentation.divio.com/structure/ defines.

----

I suspect that there is going to have to be some negoitation between pydata-sphinx-theme and sphinx-gallery to tweak how they generate TOCs to fold some things nicer.

**Comment by tacaswell** [comment]
This is where I uncritically adopted the install card from: http://scipy.github.io/devdocs/getting_started.html#getting-started-ref

File: doc/index.rst
Comment (tacaswell) [reply doc/index.rst]: Moved this to "user guide"

File: doc/index.rst
Comment (tacaswell) [reply doc/index.rst]: removed the nesting a different way.

**Comment by timhoffm** [comment]
> > I still struggle with the toc structure.
> 
> It is made worse because the `.. toc::` structure is also used by sphinx to decide things like the "next" buttons, the auto-tocs on the left and right, and the the bread-crumbs in the top-bar (for example if you put anything else from the top-bar in the user in another one, they both show bold when you are in the inner one which I found very confusing).

We should hide the prev/next buttons. They only make sense in a linear book-style doc, not in a hierarchial structure where each page stands for a dedicated topic.
Here is how to: https://pydata-sphinx-theme.readthedocs.io/en/latest/user_guide/configuring.html#hiding-the-previous-and-next-buttons

Auto-tocs should be more or less ok for us.

Top-bar is an issue. Can we work around that using external links? https://pydata-sphinx-theme.readthedocs.io/en/latest/user_guide/configuring.html#adding-external-links-to-your-nav-bar

 
> I _think_ the toc rules should be that `contents.rst` is the root of all files that have tocs and you can only use `.. toc::` in files that are in a toc in a file that is already referenced.

Yes.

> 
> My concern with your proposed layout @timhoffm is that we lose a way to pull apart the 4 kinds of docs that https://documentation.divio.com/structure/ defines.

I can't follow. The 4 kinds of docs should become elements in the "User documentation" section once we get to properly distinguish between them.

**Comment by jklymak** [comment]
Taking a quick look, I don't think its impossible to move "Plot Types" into "Gallery", and just putting them at the top.  Not sure how nestable S-G is.  

OTOH, I think I agree that "Contributing" does not need a top-level link in the banner.  I think folks who want to contribute can find that either on the landing page or the "Users Guide" or via google, 99.9% of users probably don't "Contribute", and those that do, probably only need the Contribute guide until they get their feet under them.  

Thanks, this is looking good!

**Comment by tacaswell** [comment]
> I can't follow. The 4 kinds of docs should become elements in the "User documentation" section once we get to properly distinguish between them.

Sorry, I think I miss-understood your outline with the gallery and tutorials being under "Explanation" rather than as peers to it.

> We should hide the prev/next buttons. 

done

> I don't think its impossible to move "Plot Types" into "Gallery", and just putting them at the top

For the "New to Matplotlib" user above I think it is worth having the very streamlined "this is the kind of thing you can do" entry point always easy to get to.  

> Top-bar is an issue. Can we work around that using external links?

The external links in the conf adds links with the `->` icon next to it:

image

however, I stand by not keeping the link to third-party in the top line 


I'll play around later with trying to use (internal) links in the top bar and seeing what that does, but my suspicion is that it is going to break the bread-crumb behavior.

**Comment by jklymak** [comment]
Agree about third party as a top bar issue.  Users don't need to frequently access that and I assume it is on both the landing page and somewhere in the user guide.

**Comment by tacaswell** [comment]
ok, I think I got the top bar and over all structure to be what @timhoffm proposed last week!

**Comment by timhoffm** [comment]
Thanks Tom, this looks great!

Still two things:

- The top bar "Release notes" link is broken.

- Coming back again to the Release notes toc splitting into several files (https://github.com/matplotlib/matplotlib/pull/20867#discussion_r693565344)
  I don't follow the most recent changes in the sphinx-pydata-theme but at least in CI this renders  as
  https://63285-1385122-gh.circle-artifacts.com/0/doc/build/html/users/release_notes.html and 
  https://63285-1385122-gh.circle-artifacts.com/0/doc/build/html/users/relnotes/3.4.html
  which needs annoyingly many clicks to get somewhere.
  I'd like to have this type of page back https://matplotlib.org/devdocs/users/release_notes.html.

File: doc/index.rst
Comment (anntzer) [inline doc/index.rst:46]: typo: available

**Comment by tacaswell** [comment]
> The top bar "Release notes" link is broken.

I had a commit that fixed that that I had failed to push 🤦🏻 

> Coming back again to the Release notes toc splitting into several files

dropped that commit (and opened #21027 ) to talk about that later.

**Comment by timhoffm** [comment]
Doc build failure: 

> /home/circleci/project/doc/index.rst:90: WARNING: unknown document: api/toolkits/index

File: doc/index.rst
Comment (timhoffm) [inline doc/index.rst:90]: Remove this bullet point because toolkits are part of the API index now (#21003).

**Comment by tacaswell** [comment]
I really do not like they way this renders the left nav on the release notes:

image

**Review by timhoffm — APPROVED** [review]
I agree that the left nav for the "Release notes" is awkward. Probably, because all next-lower level sections are on the page itself and I assume that the on-page sections go on the right nav whereas links to other pages go to the left nav.

Anyway, I propose to merge nonetheless. It would even be bearable to use this state for release. While the left nav is pretty useless as is, I assume that people will be visually drawn to the large "Version x.y" headings on the page and mostly ignore the noise on the left bar. Also we have #21027 to address this further.

File: doc/conf.py
Comment (jklymak) [inline doc/conf.py:297]: Did you want to keep this comment?

File: doc/index.rst
Comment (jklymak) [inline doc/index.rst:22]:
```suggestion
    ^^^^^^^^^^^^^^^^^^^^^^
```

File: doc/index.rst
Comment (jklymak) [inline doc/index.rst:35]:
```suggestion
    ^^^^^^^^^^^^^^^^^^^^
```

File: doc/index.rst
Comment (jklymak) [inline doc/index.rst:46]:
```suggestion
Further details are available in the :doc:`Installation Guide <users/installing>`.
```

File: doc/index.rst
Comment (jklymak) [inline doc/index.rst:114]:
```suggestion
Site map
```

File: doc/index.rst
Comment (jklymak) [inline doc/index.rst:25]: I don't think we need this extra verbiage:

```suggestion
```

File: doc/index.rst
Comment (jklymak) [inline doc/index.rst:21]:
```suggestion
    Installing using `conda <https://docs.continuum.io/anaconda/>`__
```

File: doc/index.rst
Comment (jklymak) [inline doc/index.rst:34]:
```suggestion
    Installing using `pip <https://pypi.org/project/matplotlib>`__
```

File: doc/index.rst
Comment (jklymak) [inline doc/index.rst:38]: As above:

```suggestion
```

File: doc/index.rst
Comment (jklymak) [inline doc/index.rst:91]: Maybe this should be a sublist because these are all under `api/index`.  But not sure how it renders...
```suggestion
      - :doc:`Axes API <api/axes_api>` for *most* plotting methods
      - :doc:`Figure API <api/figure_api>` for figure-level methods
      - Top-level interfaces to create:

        - Figures (`.pyplot.figure`)
        - Subplots (`.pyplot.subplots`, `.pyplot.subplot_mosaic`)
```

File: doc/faq/index.rst
Comment (jklymak) [inline doc/faq/index.rst:4]:
```suggestion
How-to
```

**Review by jklymak — APPROVED** [review]
A bunch of minor comments that you can take or leave and/or could be  addressed after merging

I guess this should be squashed as well, but I think anyone can merge if squashed....

File: doc/conf.py
Comment (tacaswell) [reply doc/conf.py:297]: It should go as I think it is extra-useless as this is read by `navbar_center` template that got replaced a bit further down.

File: doc/index.rst
Comment (tacaswell) [reply doc/index.rst:46]: Sorry I missed this in my last pass.

File: doc/index.rst
Comment (tacaswell) [reply doc/index.rst:91]: I'm going to hold off on this one as if everything _but_ one is in the second sub-list we should find some other way to structure this.

**Comment by tacaswell** [comment]
I squashed it down to three commits (one for the sponsor button, one for the version bump, one for everything else).

**Comment by timhoffm** [comment]
Let‘s go. 🎉  Further refinements can be done in follow-up PRs.

**Pull request merged**
